AK/UFixedBigInt.h

```diff
@@ -534,6 +535,96 @@ requires(sizeof(T) >= sizeof(u64) && IsUnsigned<T>) class UFixedBigInt {
         return res;
     }
 
+    template<Unsigned U>
+    requires(IsSame<R, U>&& IsSame<T, u64>) constexpr Tuple<R, R> wide_multiply(U const& other) const
+    {
+        auto mult_64_to_128 = [](u64 a, u64 b) -> Tuple<u64, u64> {
+#ifdef __SIZEOF_INT128__
+            unsigned __int128 result = (unsigned __int128)a * b;
+            u64 low = result;
+            u64 high = result >> 64;
+            return { low, high };
+#else
+            u32 a_low = a;
+            u32 a_high = (a >> 32);
+            u32 b_low = b;
+            u32 b_high = (b >> 32);
+
+            u64 ll = (u64)a_low * b_low;
+            u64 lh = (u64)a_low * b_high;
+            u64 hl = (u64)a_high * b_low;
+            u64 hh = (u64)a_high * b_high;
+
+            u64 lh_low = (lh << 32);
+            u64 lh_high = (lh >> 32);
+            u64 hl_low = (hl << 32);
+            u64 hl_high = (hl >> 32);
+
+            bool carry = Checked<T>::addition_would_overflow(ll, lh_low);
+            u64 low = ll + lh_low;
+            u64 high = lh_high + carry + hl_high;
+
+            carry = Checked<T>::addition_would_overflow(low, hl_low);
+            low += hl_low;
+            high += hh + carry;
+            return { low, high };
+#endif
+        };
+
+        auto ll_tuple = mult_64_to_128(m_low, other.low());
+        auto lh_tuple = mult_64_to_128(m_low, other.high());
+        auto hl_tuple = mult_64_to_128(m_high, other.low());
+        auto hh_tuple = mult_64_to_128(m_high, other.high());
+
+        R ll { ll_tuple.template get<0>(), ll_tuple.template get<1>() };
+        R llh { 0u, lh_tuple.template get<0>() };
+        R hlh { lh_tuple.template get<1>(), 0u };
+        R lhl { 0u, hl_tuple.template get<0>() };
```

Comment:
Using Tuple in non-generic library code is a bit :(. all the get<N> and such isn't easy on the eyes. Generally I'd prefer to see a simple `struct 128BitMulResult { u64 low; u64 high; }` than Tuple<R, R>.

Reply:
Yeah I don't really like the tuple either, I guess a templated `MulResult` struct might be nicer.
What I really want to do is return a `UFixedBigInt<R>`, however I could not get that to work, maybe @Hendiadyoin1 could help me with that?

Reply:
I have changed this to a struct for now, see the latest commit.